### PR TITLE
IterativeRefinement to improve EqualityConstrainedQP speed

### PR DESCRIPTION
**Pull request by Algue-Rythme**
## What is new ?

The class `IterativeRefinement` is a wrapper to increases the precision of linear system solvers: by minimizing the residuals iteratively a high accuracy solution can be obtained from a low accuracy solver, which is useful in `float32` for ill-posed problems.

This functionnality is also available through a function in `linear_solve`, that mimic the interface of `solve_gmres,solve_bicgstab` etc...

Iterative Refinement solves `Ax=b` with the following recursion:

iterative

Where `x_0=0` and `r_0=b`. The matrix `A_bar` can be any (regularized) approximation of `A`. The linear solver on the first step can be inaccurate, as long as computing the residuals `b-Ax` is accurate.

The user can specify its own `A_bar`; otherwise we take `A_bar=A`.

## Link with EqualityConstrainedQP

This allows `EqualityConstrainedQP` to solve badly scaled problems (see new test).
The following attributes have been added to EqualityConstrainedQP:

- refine_regularization: (optional) a positive float specifying the strength of regularization.
- refine_maxiter: maximum number of refinement steps when refine_regularization is not None.

## Question

Under which section should we document the class `IterativeRefinement` ? Do we need a `Linear Solvers` section ? Or under `constrained optimization` ? Finding a feasible point (constrained problem, without objective) can be turned into an optimization problem (unconstrained minimization of residuals).

### CODE REVIEW

File: jaxopt/_src/iterative_refinement.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    tol: absolute tolerance for stoping criterion (default: 1e-7).
+    verbose: If verbose=1, print error at each iteration.
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled iterations.
+    implicit_diff_solve: the linear system solver to use.
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto")
+
+  References:
+    J. H. Wilkinson. Rounding Errors in Algebraic Processes. Prentice Hall, Englewood Cliffs, NJ, 1963.
+    Moler, C.B., 1967. Iterative refinement in floating point. Journal of the ACM (JACM), 14(2), pp.316-321.
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:98]: Add link to https://en.wikipedia.org/wiki/Iterative_refinement

File: jaxopt/_src/iterative_refinement.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+             params_A_bar: Optional[Any] = None):
+    if self._copy_params_A:
+      params_A_bar = params_A
+
+    A = self.matvec_A(params_A)
+    A_bar = self.matvec_A_bar(params_A_bar)
+
+    # TODO(lbethune): support preconditioners ?
+    # Could it be done by user with partial(solver, M=precond) ?
+    residual_sol = self.solve(A_bar, state.target_residuals, x0=state.init)
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:135]: In `cg`, the option is called `init` (https://github.com/google/jaxopt/blob/main/jaxopt/_src/linear_solve.py#L91). We should probably add an `init` option to `gmres` and other solvers instead of relying on `kwargs`...

File: jaxopt/_src/iterative_refinement.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                          params_A: Any,
+                          b: Any,
+                          params_A_bar: Optional[Any] = None):
+    del params_A_bar  # unused
+    return tree_l2_norm(self.optimality_fun(params, params_A, b))
+
+  def __post_init__(self):
+    # The inner solver might not support autodiff, so 
+    # we force implicit_diff in every case.
+    self.implicit_diff = True
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:173]: Is this necessary? (https://github.com/google/jaxopt/blob/main/jaxopt/_src/base.py#L120)

File: jaxopt/_src/iterative_refinement.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  Args:
+    matvec: product between ``A`` and a vector.
+    b: pytree.
+    maxiter: maximum number of refinement steps (default: 10).
+    tol: absolute tolerance residuals (default: 1e-7).
+    solve: optional solve function (default: linear_solve.solve_gmres).
+    kwargs: additional parameters for IterativeRefinement.
+
+  Returns:
+    array with same structure as ``b``.
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:202]:
```suggestion
    Pytree with the same structure as ``b``.
```

File: jaxopt/_src/iterative_refinement.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  """
+  iterative_refinement = IterativeRefinement(matvec_A=lambda _,x: matvec(x),
+                                             solve=solve,
+                                             maxiter=maxiter,
+                                             tol=tol,
+                                             **kwargs)
+  return iterative_refinement.run(init, None, b)[0]
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:209]: Use keyword arguments (it's not clear what None refers to).

File: jaxopt/_src/linear_operator.py [hunk 6]
```diff
@@ -25,6 +25,7 @@ class DenseLinearOperator:
 
   def __init__(self, pytree):
     self.pytree = pytree
+    self.linop_type = 'dense'
```
Comment (mblondel) [inline jaxopt/_src/linear_operator.py:28]: I'm not a fan of this.

File: jaxopt/_src/linear_operator.py [hunk 7]
```diff
@@ -59,6 +60,7 @@ class FunctionalLinearOperator:
 
   def __init__(self, fun, params):
     self.fun = functools.partial(fun, params)
+    self.linop_type = 'functional'
```
Comment (mblondel) [inline jaxopt/_src/linear_operator.py:63]: Ditto

File: jaxopt/_src/osqp.py [hunk 8]
```diff
@@ -435,26 +441,34 @@ def _solve_linear_system(self, params, Q, c, A, rho_bar, state):
     x, z = params.primal
     y    = params.dual_eq  # dual variables for constraints z_bar = z;
 
-    def matvec(x_bar):
-      Qx_sigmax = tree_add_scalar_mul(Q(x_bar), self.sigma, x_bar)
-      ATAx = A.normal_matvec(x_bar)
-      return tree_add_scalar_mul(Qx_sigmax, rho_bar, ATAx)
+    if Q.linop_type == A.linop_type == 'dense':
```
Comment (mblondel) [inline jaxopt/_src/osqp.py:444]: Use `isinstance(Q, DenseLinearOperator)` instead.

File: jaxopt/_src/tree_util.py [hunk 9]
```diff
@@ -58,6 +58,10 @@ def tree_vdot(tree_x, tree_y):
   return tree_reduce(operator.add, vdots)
 
 
+def tree_dot(tree_x, tree_y):
```
Comment (mblondel) [inline jaxopt/_src/tree_util.py:61]: Add docstring (what does it mean to multiply two pytrees?)

File: tests/eq_qp_test.py [hunk 10]
```diff
@@ -140,6 +143,29 @@ def matvec_A(A, tup):
...
+
+    onp.random.seed(511)
+    X = onp.random.rand(size_prob, size_prob)
+    Q = jnp.array(X @ X.T)
+    c = jnp.array(onp.random.rand(size_prob))
+
+    A = jnp.array(onp.random.rand(size_eq, size_prob))
+    b = jnp.array(onp.random.rand(size_eq))
+    
+    # The tolerance is low, but ordinary this problem cannot be solved by
```
Comment (mblondel) [inline tests/eq_qp_test.py:158]:
```suggestion
    # The tolerance is low, but ordinarily this problem cannot be solved by
```

File: tests/iterative_refinement_test.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    sigma = 0.05
+    A_bar = A + sigma * noise  # perturbed system.
+    
+    x0 = onp.random.randn(n)  # expected solution.
+    b = A @ x0  # unperturbed target.
+
+    high_acc = 1e-3
+    solver = IterativeRefinement(matvec_A=None, matvec_A_bar=jnp.dot,
+                                 tol=high_acc, maxiter=100)
+    x, state = solver.run(None, A, b, A_bar)
```
Comment (mblondel) [inline tests/iterative_refinement_test.py:96]: Use keyword arguments.

File: tests/iterative_refinement_test.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    b = onp.random.randn(n)
+
+    low_acc = 1e-1
+    high_acc = 1e-5
+
+    # Heavily regularized low acuracy solver.
+    inner_solver = partial(_linear_solve.solve_gmres, tol=low_acc, ridge=1e-3)
+
+    solver = IterativeRefinement(solve=inner_solver, tol=high_acc, maxiter=10)
+    x, state = solver.run(None, A, b)
```
Comment (mblondel) [inline tests/iterative_refinement_test.py:46]: Use keyword arguments.

File: tests/iterative_refinement_test.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    onp.random.seed(0)
+    n = 20
+
+    A = onp.random.rand(n, n)  # invertible matrix (with high probability).
+
+    noise = onp.random.randn(n, n)
+    sigma = 0.05
+    A_bar = A + sigma * noise  # perturbed system.
+    
+    x0 = onp.random.randn(n)  # expected solution.
```
Comment (mblondel) [inline tests/iterative_refinement_test.py:90]:
```suggestion
    expected = onp.random.randn(n)
```

**Review by mblondel — CHANGES_REQUESTED** [review]
Thanks Louis!

File: jaxopt/_src/eq_qp.py [hunk 14]
```diff
@@ -62,21 +67,33 @@ def optimality_fun(params, params_obj, params_eq):
   return optimality_fun
 
 
+@dataclass
+class EqualityConstrainedQPState:
```
[the remark this answers is not in the document]
Comment (Algue-Rythme) [reply jaxopt/_src/eq_qp.py:71]: No, indeed. I am waiting for `linear_solve` solvers to return a Tuple whose second member without be infos on the algforithm (error, iter_num, etc... ). 

Not useful currently but I hope it will be in the future once linear solvers adopt the Jaxopt base.IterativeSolver interface.

File: jaxopt/_src/eq_qp.py [hunk 15]
```diff
@@ -87,13 +104,32 @@ class EqualityConstrainedQP():
...
   implicit_diff_solve: Optional[Callable] = None
   jit: bool = True
 
+  def _refined_solve(self, matvec, b, init, maxiter, tol):
+    def matvec_A(_, x):
+      return matvec(x)
+
+    ridge = self.refine_regularization
+
+    def matvec_A_bar(_, x):
```
[the remark this answers is not in the document]
Comment (Algue-Rythme) [reply jaxopt/_src/eq_qp.py:120]: I added some comments and references: it appears that quasi-definite matrices have been studied in the context of _interior point methods._ They are no stranger to quadratic programming:

Vanderbei, R.J., 1995. Symmetric quasidefinite matrices.
SIAM Journal on Optimization, 5(1), pp.100-113.

File: jaxopt/_src/eq_qp.py [hunk 16]
```diff
@@ -87,13 +104,32 @@ class EqualityConstrainedQP():
...
+      p, d = x
+      p_new, d_new = matvec(x)
```
[the remark this answers is not in the document]
Comment (Algue-Rythme) [reply jaxopt/_src/eq_qp.py:122]: You are right this variable's name is awful (and not even right): I change that !

File: jaxopt/_src/iterative_refinement.py (hunk 2 above)
Comment (Algue-Rythme) [reply jaxopt/_src/iterative_refinement.py:135]: Good call !

On the other hand the `jax.scipy.sparse.linalg` interface (that the user might be more familar with) uses `x0`, and that's what is used under the hood for `jaxopt.linear_solve.solve_gmres`.

There is a mismatch between those two interfaces.

File: jaxopt/_src/iterative_refinement.py (hunk 3 above)
Comment (Algue-Rythme) [reply jaxopt/_src/iterative_refinement.py:173]: It seems that the `run` method is wrapped only if `implicit_diff=True` but I might be wrong: https://github.com/google/jaxopt/blob/648f2f2553b1d91eaf54577ea0413ad64729089f/jaxopt/_src/base.py#L172

File: jaxopt/_src/tree_util.py (hunk 9 above)
Comment (Algue-Rythme) [reply jaxopt/_src/tree_util.py:61]: I added:

_Compute leaves-wise dot product between pytree of arrays._
  
  _Useful to store block diagonal linear operator: each leaf of the tree
  corresponds to a block._

File: jaxopt/_src/iterative_refinement.py (hunk 3 above)
Comment (mblondel) [reply jaxopt/_src/iterative_refinement.py:173]: `getattr(self, "implicit_diff", True)` returns `True` if "implicit_diff" is not present.

Comment (Algue-Rythme) [reply jaxopt/_src/iterative_refinement.py:173]: Oh yeah I didn't realized that !

**Comment by Algue-Rythme** [comment]
Fixup commit; ready for merge after final review.

File: jaxopt/_src/eq_qp.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mblondel) [reply jaxopt/_src/eq_qp.py:71]: Let's delete it for the time being then.

File: jaxopt/_src/iterative_refinement.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+  """
+  iter_num: int
+  error: float
+  target_residuals: Any
+  init: Optional[Any]
+  # TODO(lbethune): in the future return the state of the internal
+  # solver (iter_num, error) as part of the current state.
+
+
+@dataclass
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:60]:
```suggestion
@dataclass(eq=False)
```

File: jaxopt/_src/iterative_refinement.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 init_params: Any,
+                 params_A: Any,
+                 b: Any,
+                 params_A_bar: Any = None):
+    return IterativeRefinementState(
+      iter_num=0,
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:122]:
```suggestion
      iter_num=jnp.asarray(0),
```

File: jaxopt/_src/iterative_refinement.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+      error=jnp.inf,
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:123]:
```suggestion
      error=jnp.asarray(jnp.inf),
```

**Review by mblondel — CHANGES_REQUESTED** [review]
Some more comments. Let's merge after this. Also, could you squash the commits?

**Comment by Algue-Rythme** [comment]
Commits squashed + modifications made as suggested.
